modules/images/webp-uploads/helper.php

```diff
@@ -240,3 +240,19 @@ function webp_uploads_get_attachment_sources( $attachment_id, $size = 'thumbnail
 	// Return an empty array if no sources found.
 	return array();
 }
+
+/**
+ * Verify if the request if for the frontend context.
+ * The is_feed() function is to check if the query is for feeds (rdf, rss, rss2, atom).
+ *
+ * @since n.e.x.t
+ *
+ * @return bool
+ */
+function webp_uploads_is_frontend_context() {
+	if ( ! did_action( 'template_redirect' ) || doing_action( 'wp_head' ) || doing_action( 'wp_footer' ) || is_feed() ) {
```

Comment:
I don't think we should include `wp_footer` here. It can also be used to inject stuff that is shown to the user, and most importantly, it _is_ within the `body` tag which is generally what's used for user-facing content. You're missing the check for `! did_action( 'wp_head' )` though - without that any code that executes after `template_redirect` but before `wp_head` would get `true` here which wouldn't be correct.

`is_feed()` isn't absolutely necessary since the other conditions would already cater for that, however I like the idea of including some of those cheaper checks here to not have to call the action checks unless necessary. In that case though, I think we should also include something for a REST API request, and we should put those first, above the hook checks.

We also need to check whether the `$wp_query` global is set, otherwise `is_feed()` can cause PHP warnings. In addition, it is safe to say that if that object is not set, we're not in a frontend context.

```suggestion
	global $wp_query;

	// Check if this request is generally outside (or before) any frontend context.
	if ( ! isset( $wp_query ) || defined( 'XMLRPC_REQUEST' ) || defined( 'REST_REQUEST' ) || is_feed() ) {
		return false;
	}

	// Check if we're anywhere before the 'wp_head' action has completed.
	if ( ! did_action( 'template_redirect' ) || ! did_action( 'wp_head' ) || doing_action( 'wp_head' ) ) {
```

Reply:
@felixarntz In my test results, with the default environment settings, images were not replaced in the content as well when I added the check for `! did_action( 'wp_head` )`

Reply:
@akshitsethi 
> @felixarntz In my test results, with the default environment settings, images were not replaced in the content as well when I added the check for `! did_action( 'wp_head` )`

Hmm, I can't imagine that - can you provide some more detail? `wp_head` runs in the `<head>` of the page where no user-facing content is rendered, so we have to make sure this hook has already run. For example if someone includes an image _before_ `wp_head`, it should not be replaced.